src/crimson/osd/replicated_backend.cc

```diff
@@ -17,3 +31,61 @@ seastar::future<bufferlist> ReplicatedBackend::_read(const hobject_t& hoid,
 {
   return store->read(coll, ghobject_t{hoid}, off, len, flags);
 }
+
+seastar::future<ceph::osd::acked_peers_t>
+ReplicatedBackend::_submit_transaction(std::set<pg_shard_t>&& pg_shards,
+                                       const hobject_t& hoid,
+                                       ceph::os::Transaction&& txn,
+                                       osd_reqid_t req_id,
+                                       epoch_t min_epoch, epoch_t map_epoch,
+                                       eversion_t ver)
+{
+  const ceph_tid_t tid = next_txn_id++;
+  auto pending_txn = pending_trans.emplace(tid, pending_on_t{}).first;
+  bufferlist encoded_txn;
+  encode(txn, encoded_txn);
+
+  return seastar::parallel_for_each(std::move(pg_shards),
```

Comment:
as discussed over ceph-devel mailing list, i planned to expose the underlying blockers in a finer granularity by adding a customized blocker which can dump the acked peer and if the local storage has committed the transaction. the change is stashed at https://github.com/tchaikov/ceph/tree/wip-crimson-rep-op-gate

the reason why i want to drop the blocker change from this PR is that,

1. i want to understand the performance impact of an extra layer of abstraction
2. i want to have more inputs on if we want to track multiple parallel pending transactions using blocker machinery, if yes, what's the granularity we want to have? as, after all shards commits the transaction, we still need to reply to the client. and to send the reply is also a blocking call. shall we track this as well? if not, `ClientRequest::PGPipeline::process` will be identical to ("prepare transaction + commit transaction" | "perform read ops").

quote from @rzarzynski :

>      // TODO: issue requests in parallel if they don't write,
>      // with writes being basically a synchronization barrier

so, a more interesting problem is, for better read performance, shall we try to grab as many read-only ops as possible from `MOSDOp::ops` and submit them in parallel before hitting an op which has side-effects? if yes, we can repeat this step until we reach the last ops in `MOSDOp::ops`. and then wait until the last batch of read promises are fulfilled or the last write transaction is committed, then get back to the client.

Reply:
as per the discussion with @athanatos over the standup, in the case of RBD client, it's unlikely it issues command to different snaps of a certain object in a single MOSDOp. so there is no need to parallelize the execution. this also applies to the RGW and cephfs clients.